File: src/main/java/com/vc/hard/L1063.scala
package com.vc.hard

object L1063 {
  def validSubarrays(arr: Array[Int]): Int = {
    /**
          [1,4,2,5,3]
          Next smaller element
      */
    import scala.collection.mutable
    val st = new mutable.Stack[Int]()
    var res = 0
    val n = arr.length
    var i = n - 1
    while(i >= 0) {
      while(st.nonEmpty && arr(i) <= arr(st.head)) st.pop
      res += (if(st.nonEmpty) st.head - i else n - i)
      st.push(i)
      i -= 1
    }
    res
  }
}

File: src/main/java/com/vc/hard/L1063.java
package com.vc.hard;

import java.util.*;

class L1063 {
    public int validSubarrays(int[] arr) {
        Stack<int[]> st = new Stack<>();
        int ans = 0;
        for(int i = arr.length - 1; i >= 0; i--) {
            int count = 0;
            while(!st.isEmpty() && st.peek()[0] >= arr[i]) {
                int[] e = st.pop();
                count += e[1] + 1;
            }
            ans += count + 1;
            st.push(new int[]{arr[i], count});
        }
        return ans;
    }
}
